File: apps/windows/TeslaSync.App/dashboard-widgets/BatteryDegradationTrendWidget/BatteryDegradationTrendWidget.Model.cs
using System.Globalization;
using System.Text.Json;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.DashboardWidgets;

/// <summary>
/// The lifecycle state a <see cref="BatteryDegradationTrendViewModel"/> can be in — the native union of the
/// loading / loaded / empty / error / stale / offline branches the web <c>BatteryDegradationTrendWidget</c>
/// renders through <c>WidgetShell</c> + <c>WidgetChartSummary</c>
/// (web/src/features/dashboard/widgets/BatteryDegradationTrendWidget.tsx). Every branch maps onto a visible
/// surface; none is ever hidden. <see cref="Empty"/> mirrors the web <c>isEmpty</c> gate (no current-health
/// value AND no monthly-trend rows) — the friendly "No degradation data" empty state — distinct from a
/// transport failure (<see cref="Error"/>).
/// </summary>
public enum BatteryDegradationTrendState
{
    /// <summary>Initial fetch with no cached trend — render the skeleton chrome.</summary>
    Loading,

    /// <summary>A fresh trend (or non-stale cache) carrying a current-health value or trend rows.</summary>
    Loaded,

    /// <summary>No vehicle resolved, or a trend with no usable data — render the empty state.</summary>
    Empty,

    /// <summary>The request failed and no cached trend exists — render the retry affordance.</summary>
    Error,

    /// <summary>A cached trend older than the freshness window — render content plus a stale chip.</summary>
    Stale,

    /// <summary>The network failed but a cached trend remains — render content plus an offline chip.</summary>
    Offline,
}

/// <summary>
/// One monthly aggregation row from <c>GET /analytics/battery-degradation</c> (web <c>DegradationTrend</c>
/// in web/src/types/energy.ts): the <see cref="Month"/> bucket label, the average state-of-health
/// percentage (<c>avg_health</c>, the value the web area chart plots), and the average rated range
/// (<c>avg_range</c>). Field names mirror the Go API's snake_case JSON tags; parsing is null-tolerant so a
/// partial row never throws.
/// </summary>
public sealed record DegradationTrendPoint(string Month, double AvgHealth, double AvgRange)
{
    /// <summary>Project a single monthly-trend JSON object into a tolerant <see cref="DegradationTrendPoint"/>.</summary>
    public static DegradationTrendPoint FromJson(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return new DegradationTrendPoint(string.Empty, 0, 0);
        }

        return new DegradationTrendPoint(
            Month: TrendJson.String(element, "month") ?? string.Empty,
            AvgHealth: TrendJson.Double(element, "avg_health") ?? 0,
            AvgRange: TrendJson.Double(element, "avg_range") ?? 0);
    }
}

/// <summary>
/// The battery-degradation trend read-model the widget consumes — the subset of the
/// <c>GET /analytics/battery-degradation</c> body (web <c>DegradationData</c>) the web
/// <c>BatteryDegradationTrendWidget</c> reads: the current health percentage
/// (<c>current_health_pct ?? current_health</c>), the monthly degradation rate, the lifetime cycle count,
/// and the monthly health/range trend. All percentages are already display-ready (the web shows them raw,
/// no unit conversion). Parsing is tolerant so a partial or non-object body yields <see cref="Empty"/>
/// rather than throwing.
/// </summary>
public sealed record BatteryDegradationTrend(
    double? CurrentHealthPct,
    double? DegradationRatePctPerMonth,
    double? CurrentCycles,
    IReadOnlyList<DegradationTrendPoint> MonthlyTrend)
{
    /// <summary>A data-free trend — the parse fallback for an absent/non-object body.</summary>
    public static BatteryDegradationTrend Empty { get; } =
        new(null, null, null, Array.Empty<DegradationTrendPoint>());

    /// <summary>
    /// True when the whole surface has nothing to show (web <c>isEmpty = currentHealth == null &amp;&amp;
    /// chartData.length === 0</c>): no resolved current-health value and no monthly-trend rows.
    /// </summary>
    public bool IsEmpty => CurrentHealthPct is null && MonthlyTrend.Count == 0;

    /// <summary>Project a <c>GET /analytics/battery-degradation</c> JSON body into a tolerant trend.</summary>
    public static BatteryDegradationTrend FromJson(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return Empty;
        }

        // Web parity: current_health_pct wins, then current_health, then null (the ?? chain only falls
        // through on absent/null — a literal 0 is a valid health value and is kept).
        double? currentHealth = TrendJson.Double(element, "current_health_pct")
            ?? TrendJson.Double(element, "current_health");

        return new BatteryDegradationTrend(
            CurrentHealthPct: currentHealth,
            DegradationRatePctPerMonth: TrendJson.Double(element, "degradation_rate_pct_per_month"),
            CurrentCycles: TrendJson.Double(element, "current_cycles"),
            MonthlyTrend: ReadTrend(element));
    }

    private static IReadOnlyList<DegradationTrendPoint> ReadTrend(JsonElement element)
    {
        if (!element.TryGetProperty("monthly_trend", out var arr) || arr.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<DegradationTrendPoint>();
        }

        var list = new List<DegradationTrendPoint>(arr.GetArrayLength());
        foreach (var item in arr.EnumerateArray())
        {
            list.Add(DegradationTrendPoint.FromJson(item));
        }

        return list;
    }
}

/// <summary>
/// The widget's grid footprint (columns × rows). Mirrors the web <c>WidgetProps.size</c> and the
/// <c>isCompact = size.cols &lt;= 1 &amp;&amp; size.rows &lt;= 1</c> branch in
/// web/src/features/dashboard/widgets/BatteryDegradationTrendWidget.tsx.
/// </summary>
public readonly record struct BatteryDegradationTrendSize(int Cols, int Rows)
{
    /// <summary>The registry default footprint (2×4).</summary>
    public static BatteryDegradationTrendSize Default => new(2, 4);

    /// <summary>True at a single cell (web <c>isCompact</c>): hide the title and chart, show the stats only.</summary>
    public bool IsCompact => Cols <= 1 && Rows <= 1;
}

/// <summary>
/// One projected, display-ready stat from the summary row — the native analogue of a web
/// <c>ChartSummaryStat</c>. Holds the localized <see cref="Label"/>, the formatted <see cref="Value"/>, the
/// optional <see cref="Unit"/> suffix (e.g. <c>/mo</c>), and a Narrator automation name. Pure data — no
/// WinUI types.
/// </summary>
public sealed record TrendSummaryStat(string Label, string Value, string? Unit, string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the degradation trend for one footprint — the native analogue
/// of everything the web component computes via <c>useMemo</c> before returning JSX. Holds the summary
/// stats, the area-chart health series (with its month labels and end-of-life reference flag), and the
/// localized chart copy. Pure data so the projection is unit-tested without a UI host.
/// </summary>
public sealed record BatteryDegradationTrendDisplay(
    bool IsCompact,
    bool IsEmpty,
    IReadOnlyList<TrendSummaryStat> Stats,
    bool HasChart,
    IReadOnlyList<double> ChartHealth,
    IReadOnlyList<string> ChartMonths,
    bool ShowEolThreshold,
    string ChartSeriesName,
    string ChartEmptyMessage,
    string CompactAutomationName);

/// <summary>
/// Pure projection from a raw <see cref="BatteryDegradationTrend"/> to the display model — the native port
/// of the <c>stats</c> / <c>chartData</c> <c>useMemo</c> work and the <c>isCompact</c> / <c>isEmpty</c>
/// gating in web/src/features/dashboard/widgets/BatteryDegradationTrendWidget.tsx. Percentages are already
/// display-ready (the web shows them raw, no unit conversion), so this only formats, labels and derives the
/// chart series; every label resolves through the i18n facade.
/// </summary>
public static class BatteryDegradationTrendProjection
{
    /// <summary>Segoe Fluent "MarketDown" glyph for the surface header / empty state (web <c>TrendingDown</c>).</summary>
    public const string HeaderGlyph = "\uEB0F";

    /// <summary>End-of-life reference percentage the chart marks with a dashed line (web <c>ReferenceLine y=80</c>).</summary>
    public const double EolThresholdPct = 80.0;

    /// <summary>Domain padding (web Y axis <c>domain={['dataMin - 2', 100]}</c>) used for the reference-line visibility test.</summary>
    public const double DomainPaddingPct = 2.0;

    private const string EmDash = "\u2014";
    private const string MinusSign = "\u2212";

    /// <summary>Project <paramref name="trend"/> for <paramref name="size"/> using the localizer for every label.</summary>
    public static BatteryDegradationTrendDisplay Project(
        BatteryDegradationTrend trend,
        BatteryDegradationTrendSize size,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(trend);
        ArgumentNullException.ThrowIfNull(localizer);

        var stats = BuildStats(trend, localizer);

        var health = new List<double>(trend.MonthlyTrend.Count);
        var months = new List<string>(trend.MonthlyTrend.Count);
        foreach (var point in trend.MonthlyTrend)
        {
            health.Add(point.AvgHealth);
            months.Add(point.Month);
        }

        // Web parity: the area chart only renders with more than one point; otherwise the "need more data"
        // message takes its place.
        bool hasChart = health.Count > 1;

        // Web parity: the y-axis domain is [dataMin - 2, 100], so the 80% reference line is only visible when
        // the lowest health sample sits at or below 82 — otherwise it falls outside the zoomed-in domain.
        bool showEol = hasChart && MinOf(health) <= EolThresholdPct + DomainPaddingPct;

        return new BatteryDegradationTrendDisplay(
            IsCompact: size.IsCompact,
            IsEmpty: trend.IsEmpty,
            Stats: stats,
            HasChart: hasChart,
            ChartHealth: health,
            ChartMonths: months,
            ShowEolThreshold: showEol,
            ChartSeriesName: localizer.GetString("widget.healthPct", "Health %"),
            ChartEmptyMessage: localizer.GetString("widget.needMoreData", "More data needed for trend"),
            CompactAutomationName: string.Join(", ", AutomationParts(stats)));
    }

    private static List<TrendSummaryStat> BuildStats(BatteryDegradationTrend trend, ILocalizer localizer)
    {
        var stats = new List<TrendSummaryStat>(3);

        string sohLabel = localizer.GetString("widget.soh", "SoH");
        string sohValue = trend.CurrentHealthPct is { } health ? $"{Fmt(health, 1)}%" : EmDash;
        stats.Add(new TrendSummaryStat(sohLabel, sohValue, null, $"{sohLabel}: {sohValue}"));

        // Web parity: the degradation stat only appears when a positive monthly rate is present.
        if (trend.DegradationRatePctPerMonth is { } rate && rate > 0)
        {
            string degLabel = localizer.GetString("widget.degradation", "Degradation");
            string degValue = $"{MinusSign}{Fmt(rate, 2)}%";
            string degUnit = $"/{localizer.GetString("widget.mo", "mo")}";
            stats.Add(new TrendSummaryStat(degLabel, degValue, degUnit, $"{degLabel}: {degValue}{degUnit}"));
        }

        string cyclesLabel = localizer.GetString("widget.cycles", "Cycles");
        string cyclesValue = trend.CurrentCycles is { } cycles ? Fmt(cycles, 0) : EmDash;
        stats.Add(new TrendSummaryStat(cyclesLabel, cyclesValue, null, $"{cyclesLabel}: {cyclesValue}"));

        return stats;
    }

    private static IEnumerable<string> AutomationParts(IReadOnlyList<TrendSummaryStat> stats)
    {
        foreach (var stat in stats)
        {
            yield return stat.AutomationName;
        }
    }

    private static double MinOf(IReadOnlyList<double> values)
    {
        double min = double.PositiveInfinity;
        foreach (double v in values)
        {
            if (v < min)
            {
                min = v;
            }
        }

        return double.IsPositiveInfinity(min) ? 0 : min;
    }

    /// <summary>
    /// Format a number exactly as the web <c>fmtNumber</c> does: coerce null / NaN / ±∞ to 0 (web
    /// <c>safeNumber</c>) then render with fixed <paramref name="decimals"/> fraction digits and en-US grouping.
    /// </summary>
    private static string Fmt(double value, int decimals)
    {
        double safe = !double.IsNaN(value) && !double.IsInfinity(value) ? value : 0.0;
        return ScalarFormatters.FormatNumber(safe, decimals);
    }
}

/// <summary>
/// Maps the engine's raw <c>RepositoryResult&lt;JsonElement&gt;</c> emissions onto parsed
/// <c>RepositoryResult&lt;BatteryDegradationTrend&gt;</c>, preserving every freshness flag
/// (cached / refreshing / stale / offline) so the view-model can render the full state matrix. The
/// <c>isEmpty</c> gate (web's <c>WidgetChartSummary isEmpty</c>) is applied by the view-model, not here, so
/// a populated-but-data-free body still flows through with its freshness intact. Kept pure so the
/// parse-and-preserve contract is unit-tested without a network or cache.
/// </summary>
public static class BatteryDegradationTrendResultMapper
{
    /// <summary>Parse <paramref name="raw"/>'s payload (when present) while preserving its status.</summary>
    public static RepositoryResult<BatteryDegradationTrend> Map(RepositoryResult<JsonElement> raw)
    {
        ArgumentNullException.ThrowIfNull(raw);

        BatteryDegradationTrend Parse() =>
            raw.HasValue ? BatteryDegradationTrend.FromJson(raw.Value) : BatteryDegradationTrend.Empty;

        return raw.Status switch
        {
            LoadStatus.Loading => RepositoryResult<BatteryDegradationTrend>.Loading(),
            LoadStatus.Cached => RepositoryResult<BatteryDegradationTrend>.Cached(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Refreshing => RepositoryResult<BatteryDegradationTrend>.Refreshing(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Loaded => RepositoryResult<BatteryDegradationTrend>.Loaded(Parse(), raw.FetchedAt ?? DateTimeOffset.UtcNow),
            LoadStatus.Empty => RepositoryResult<BatteryDegradationTrend>.Empty(raw.FetchedAt),
            LoadStatus.Offline => RepositoryResult<BatteryDegradationTrend>.OfflineCached(Parse(), raw.FetchedAt!.Value, raw.Error!),
            _ => RepositoryResult<BatteryDegradationTrend>.Failure(
                raw.Error ?? new RepositoryError(RepositoryErrorKind.Unknown, "Unknown error")),
        };
    }
}

/// <summary>
/// Null-tolerant readers for the snake_case battery-degradation JSON wire shape: numbers (or numeric
/// strings) and strings. Kept private to this surface so the trend adapter stays self-contained.
/// </summary>
internal static class TrendJson
{
    /// <summary>Reads a numeric (or numeric-string) property, or null when absent / non-numeric.</summary>
    public static double? Double(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v))
        {
            return null;
        }

        return v.ValueKind switch
        {
            JsonValueKind.Number when v.TryGetDouble(out var n) => n,
            JsonValueKind.String when double.TryParse(v.GetString(), NumberStyles.Float, CultureInfo.InvariantCulture, out var n) => n,
            _ => null,
        };
    }

    /// <summary>Reads a non-empty string property, or null when absent / non-string / blank.</summary>
    public static string? String(JsonElement obj, string name)
    {
        if (obj.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String)
        {
            string? s = v.GetString();
            return string.IsNullOrEmpty(s) ? null : s;
        }

        return null;
    }
}

File: apps/android/app/src/main/java/io/teslasync/android/dashboardwidgets/BatteryDegradationTrendWidgetModel.kt
package io.teslasync.android.dashboardwidgets

import io.teslasync.android.components.charts.ChartFormat
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import java.util.Locale

/*
 * Framework-free domain + projection for the BatteryDegradationTrend dashboard widget — the native
 * port of the data the web `BatteryDegradationTrendWidget`
 * (web/src/features/dashboard/widgets/BatteryDegradationTrendWidget.tsx) computes before it renders
 * JSX. Pure Kotlin (no Android, no Compose, no coroutines) so the tolerant JSON parsing, the
 * `current_health_pct ?? current_health` fallback, the `chartData` map, the `stats` `useMemo` and the
 * compact / empty / "needs more data" branches are all unit-tested off device. The only dependency is
 * the framework-free [ChartFormat] number formatter (the Android counterpart of the web `fmtNumber`).
 */

/**
 * One monthly degradation sample from `GET /analytics/battery-degradation` (web `DegradationTrend`).
 * Only the fields the widget reads are projected — the [month] bucket label, the average state of
 * health [avgHealth] (percent) and the average rated range [avgRange]. Reads are null-tolerant so a
 * partial row never throws (web treats them as plain numbers).
 */
data class DegradationTrend(
    val month: String,
    val avgHealth: Double?,
    val avgRange: Double?,
)

/**
 * One render-ready chart row — the native analogue of the flat objects the web builds in its
 * `chartData` `useMemo`: the [month] label, the rated [range], the [health] plotted by the area
 * series, and the [original] (the first sample's range, repeated on every row exactly as the web
 * does). Only [month] + [health] feed the native area chart; [range]/[original] are carried for
 * faithful transform parity (and the accessible data table).
 */
data class DegradationChartPoint(
    val month: String,
    val range: Double?,
    val health: Double?,
    val original: Double?,
)

/**
 * The widget's grid footprint (columns × rows). Mirrors the web `WidgetProps.size` plus the
 * `isCompact` rule, which for this widget is BOTH dimensions at one (`size.cols <= 1 && size.rows <=
 * 1`) — so the registry minimum (1×2) is NOT compact and still shows the chart, while only a 1×1 tile
 * collapses to the stat row.
 */
data class BatteryDegradationSize(
    val cols: Int,
    val rows: Int,
) {
    /** True only at a single cell (web `isCompact = size.cols <= 1 && size.rows <= 1`). */
    val isCompact: Boolean get() = cols <= COMPACT_MAX_COLS && rows <= COMPACT_MAX_ROWS

    companion object {
        /** A footprint of one column or fewer is a compact candidate (web `size.cols <= 1`). */
        const val COMPACT_MAX_COLS: Int = 1

        /** Combined with [COMPACT_MAX_COLS]: only a single-row, single-column tile is compact. */
        const val COMPACT_MAX_ROWS: Int = 1
    }
}

/**
 * The parsed `GET /analytics/battery-degradation` payload backing the widget (web `DegradationData`),
 * reduced to exactly what the surface renders: the predictive [currentHealthPct] with the legacy
 * [currentHealth] fallback, the per-month [degradationRatePctPerMonth], the lifetime [currentCycles],
 * and the [monthlyTrend]. [hasData] distinguishes a fetched payload (even an all-null one) from the
 * absent-body fallback used before anything has loaded.
 *
 * Health is read SI-free (already a percent in this analytics shape); no unit conversion happens here.
 */
data class BatteryDegradationSnapshot(
    val hasData: Boolean,
    val currentHealthPct: Double?,
    val currentHealth: Double?,
    val degradationRatePctPerMonth: Double?,
    val currentCycles: Double?,
    val monthlyTrend: List<DegradationTrend>,
) {
    /** State of health to surface: the predictive percent, falling back to the legacy field (web `??`). */
    val effectiveHealth: Double? get() = currentHealthPct ?: currentHealth

    /**
     * True when there is nothing meaningful to show — no state of health AND no trend (web
     * `isEmpty = currentHealth == null && chartData.length === 0`). Drives the [io.teslasync.android.data.UiPhase.Empty]
     * surface.
     */
    val isEmpty: Boolean get() = effectiveHealth == null && monthlyTrend.isEmpty()

    companion object {
        /** The absent-body fallback (nothing loaded yet) — flagged [hasData] = false. */
        val EMPTY: BatteryDegradationSnapshot =
            BatteryDegradationSnapshot(
                hasData = false,
                currentHealthPct = null,
                currentHealth = null,
                degradationRatePctPerMonth = null,
                currentCycles = null,
                monthlyTrend = emptyList(),
            )

        /** A fetched payload that resolved no degradation analytics (web `data == null` short-circuit). */
        val NO_DATA: BatteryDegradationSnapshot = EMPTY.copy(hasData = true)

        /** Parses a `/analytics/battery-degradation` body into a tolerant snapshot (web destructuring + `safeArray`). */
        fun fromJson(element: JsonElement?): BatteryDegradationSnapshot {
            val obj = element as? JsonObject ?: return NO_DATA
            return BatteryDegradationSnapshot(
                hasData = true,
                currentHealthPct = obj.doubleValue("current_health_pct"),
                currentHealth = obj.doubleValue("current_health"),
                degradationRatePctPerMonth = obj.doubleValue("degradation_rate_pct_per_month"),
                currentCycles = obj.doubleValue("current_cycles"),
                monthlyTrend = parseTrend(obj["monthly_trend"]),
            )
        }

        /** Project a `monthly_trend` JSON array into a tolerant list of [DegradationTrend] (web `safeArray`). */
        fun parseTrend(element: JsonElement?): List<DegradationTrend> =
            (element as? JsonArray)
                ?.mapNotNull { item -> (item as? JsonObject)?.toTrend() }
                ?: emptyList()

        private fun JsonObject.toTrend(): DegradationTrend? {
            val month = stringValue("month") ?: return null
            return DegradationTrend(
                month = month,
                avgHealth = doubleValue("avg_health"),
                avgRange = doubleValue("avg_range"),
            )
        }

        private fun JsonObject.doubleValue(key: String): Double? = (this[key] as? JsonPrimitive)?.doubleOrNull

        private fun JsonObject.stringValue(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content
    }
}

/** One summary stat chip — the native counterpart of the web `ChartSummaryStat` ({label, value, unit?}). */
data class BatteryDegradationStat(
    val label: String,
    val value: String,
    val unit: String? = null,
)

/** Localized labels for the summary stats + axis, resolved by the view from the P1/S10 i18n catalog. */
data class BatteryDegradationLabels(
    val soh: String,
    val degradation: String,
    val cycles: String,
    val perMonth: String,
)

/**
 * The fully projected, render-ready view of the degradation data for one footprint — the native
 * analogue of everything the web component computes before returning JSX: the compact branch, the
 * empty branch, the resolved [currentHealth] / [degradationRate] / [totalCycles], and the
 * newest-untouched-order chart [points]. Pure data so it is unit-tested directly; the derived flags
 * keep the type small.
 */
data class BatteryDegradationDisplay(
    val isCompact: Boolean,
    val isEmpty: Boolean,
    val points: List<DegradationChartPoint>,
    val currentHealth: Double?,
    val degradationRate: Double?,
    val totalCycles: Double?,
) {
    /** Whether the area chart is shown — web renders it only with more than one sample (`chartData.length > 1`). */
    val hasTrend: Boolean get() = points.size > 1

    /** Whether the degradation chip is shown — web pushes it only when the rate is present and positive. */
    val showDegradation: Boolean get() = degradationRate != null && degradationRate > 0.0

    /** The x-axis month labels for the chart, in source order. */
    val monthLabels: List<String> get() = points.map { it.month }

    /** The plotted health series (percent), in source order; nulls bridge gaps at the chart layer. */
    val healthValues: List<Double?> get() = points.map { it.health }
}

/**
 * Pure projection from a parsed [BatteryDegradationSnapshot] to the [BatteryDegradationDisplay] and the
 * summary stat chips — the native port of the web `chartData` + `stats` `useMemo` work plus the
 * compact branch. Percentages are formatted via the framework-free [ChartFormat] (web `fmtNumber`); the
 * em dash / minus-sign / `%` punctuation matches the web verbatim.
 */
object BatteryDegradationProjection {
    /** Decimal places per stat, matching the web `fmtNumber(value, n)` calls. */
    private const val HEALTH_DECIMALS: Int = 1
    private const val RATE_DECIMALS: Int = 2
    private const val CYCLES_DECIMALS: Int = 0

    /** U+2212 MINUS SIGN — the exact glyph the web prefixes the degradation rate with. */
    private const val MINUS_SIGN: String = "\u2212"
    private const val PERCENT: String = "%"

    /** The 80% rated-capacity warranty threshold the web draws as a reference line (see the view a11y note). */
    const val WARRANTY_THRESHOLD_PCT: Double = 80.0

    /** Project [snapshot] for [size] into the render-ready display (web `chartData` + `isCompact` + `isEmpty`). */
    fun project(
        snapshot: BatteryDegradationSnapshot,
        size: BatteryDegradationSize,
    ): BatteryDegradationDisplay {
        val trend = snapshot.monthlyTrend
        val originalRange = trend.firstOrNull()?.avgRange
        val points =
            trend.map { entry ->
                DegradationChartPoint(
                    month = entry.month,
                    range = entry.avgRange,
                    health = entry.avgHealth,
                    original = originalRange,
                )
            }
        return BatteryDegradationDisplay(
            isCompact = size.isCompact,
            isEmpty = snapshot.isEmpty,
            points = points,
            currentHealth = snapshot.effectiveHealth,
            degradationRate = snapshot.degradationRatePctPerMonth,
            totalCycles = snapshot.currentCycles,
        )
    }

    /**
     * Builds the summary stat chips exactly as the web `stats` `useMemo` does: always a SoH chip and a
     * Cycles chip (em dash when absent), with a Degradation chip inserted between them only when the
     * rate is present and positive. [labels] are resolved through the i18n facade; [locale] keeps the
     * number formatting deterministic in tests.
     */
    fun stats(
        display: BatteryDegradationDisplay,
        labels: BatteryDegradationLabels,
        locale: Locale = Locale.getDefault(),
    ): List<BatteryDegradationStat> =
        buildList {
            add(BatteryDegradationStat(label = labels.soh, value = sohValue(display.currentHealth, locale)))
            if (display.showDegradation) {
                add(
                    BatteryDegradationStat(
                        label = labels.degradation,
                        value = degradationValue(requireNotNull(display.degradationRate), locale),
                        unit = "/${labels.perMonth}",
                    ),
                )
            }
            add(BatteryDegradationStat(label = labels.cycles, value = cyclesValue(display.totalCycles, locale)))
        }

    /** SoH chip value: `{health,1}%`, or an em dash when health is absent (web `… ? `${…}%` : '—'`). */
    fun sohValue(
        health: Double?,
        locale: Locale = Locale.getDefault(),
    ): String = if (health == null) ChartFormat.EMPTY else "${ChartFormat.number(health, HEALTH_DECIMALS, locale)}$PERCENT"

    /** Degradation chip value: `−{rate,2}%` (web prefixes a U+2212 minus). Only called when the rate is positive. */
    fun degradationValue(
        rate: Double,
        locale: Locale = Locale.getDefault(),
    ): String = "$MINUS_SIGN${ChartFormat.number(rate, RATE_DECIMALS, locale)}$PERCENT"

    /** Cycles chip value: `{cycles,0}`, or an em dash when absent (web `… ? fmtNumber(…,0) : '—'`). */
    fun cyclesValue(
        cycles: Double?,
        locale: Locale = Locale.getDefault(),
    ): String = if (cycles == null) ChartFormat.EMPTY else ChartFormat.number(cycles, CYCLES_DECIMALS, locale)
}

/**
 * Static registry metadata for the BatteryDegradationTrend surface — the canonical id, category and
 * grid-size constraints from web/src/features/dashboard/widgets/registry/battery.ts. A dashboard host
 * registers the surface with this id and honors these size bounds, mirroring the web registry exactly.
 */
object BatteryDegradationTrendWidgetDescriptor {
    /** Canonical registry id (web `battery-degradation-trend`). */
    const val ID: String = "battery-degradation-trend"

    /** Registry category (web `battery`). */
    const val CATEGORY: String = "battery"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SURFACE_SLUG: String = "BatteryDegradationTrendWidget"

    /** Registry default footprint (2×4). */
    val defaultSize: BatteryDegradationSize = BatteryDegradationSize(cols = 2, rows = 4)

    /** Registry minimum footprint (1×2). */
    val minSize: BatteryDegradationSize = BatteryDegradationSize(cols = 1, rows = 2)

    /** Registry maximum footprint (4×40). */
    val maxSize: BatteryDegradationSize = BatteryDegradationSize(cols = 4, rows = 40)
}
